src/level/level_grid.rs

```diff
@@ -1,173 +1,151 @@
-        if x < 0 || y < 0 || x >= self.grid.len() as i32 {
-            return None;
-        }
-
-        self.grid
-            .get(x as usize)
-            .unwrap()
-            .get(y as usize)
-            .map(|entity| *entity)
-    }
-}
-
-impl Default for LevelGrid {
-    fn default() -> LevelGrid {
-        LevelGrid { grid: vec![vec![]] }
-    }
-}
+const CONCEALED: &str = "concealed";
+
+use amethyst::core::cgmath::{Deg, Vector3};
+use amethyst::core::transform::{GlobalTransform, Transform};
+use amethyst::ecs::prelude::Entity;
+use amethyst::ecs::storage::{GenericReadStorage, GenericWriteStorage};
+use amethyst::prelude::*;
+use entities::Tile;
+use util;
+
+pub struct LevelGrid {
+    grid: Vec<Vec<Entity>>,
+}
+
+impl LevelGrid {
+    pub fn from_grid(mut tile_grid: Vec<Vec<Tile>>, world: &mut World) -> LevelGrid {
+        let level_grid: Vec<Vec<Entity>> = tile_grid
+            .iter_mut()
+            .map(|tile_vec| {
+                tile_vec
+                    .iter_mut()
+                    .map(|tile| {
+                        world
+                            .create_entity()
+                            .with(*tile)
+                            .with(GlobalTransform::default())
+                            .build()
+                    })
+                    .collect()
+            })
+            .collect();
+
+        LevelGrid { grid: level_grid }
+    }
+
+    pub fn set_transform<T: GenericWriteStorage<Component = Transform>>(
+        &self,
+        x: i32,
+        y: i32,
+        rotation: i32,
+        mut transforms: T,
+    ) {
+        let mut transform = Transform::default();
+        transform.set_position(Vector3 {
+            x: x as f32,
+            y: 0.0,
+            z: y as f32,
+        });
+        transform.rotate_local(Vector3::new(0.0, 1.0, 0.0), Deg(-rotation as f32));
+        let entity = self.get(x as i32, y as i32).unwrap();
+        transforms.insert(entity, transform).unwrap();
+    }
+
+    pub fn direct_neighbors(&self, x: i32, y: i32) -> Vec<Entity> {
+        let mut result = Vec::with_capacity(4);
+
+        for (d_x, d_y) in [(0, 1), (0, -1), (1, 0), (-1, 0)].iter() {
+            if let Some(entity) = self.get(x + d_x, y + d_y) {
+                result.push(entity);
+            }
+        }
+        result
+    }
+
+    pub fn diagonal_neighbors(&self, x: i32, y: i32) -> Vec<Entity> {
+        let mut result = Vec::with_capacity(4);
+
+        for (d_x, d_y) in [(1, -1), (1, 1), (-1, 1), (-1, -1)].iter() {
+            if let Some(entity) = self.get(x + d_x, y + d_y) {
+                result.push(entity);
+            }
+        }
+        result
+    }
+
+    pub fn grid(&self) -> &Vec<Vec<Entity>> {
+        &self.grid
+    }
+
+    pub fn determine_sprite_for<'a, T: GenericReadStorage<Component = Tile>>(
+        &self,
+        x: i32,
+        y: i32,
+        dictionary: &'a Vec<([[Tile; 3]; 3], String)>,
```

Comment:
we need a better name than `dictionary`. since you said it maps the pattern to a mesh name, i would call it `pattern_to_mesh_name`, or perhaps `tile_mesh_names`. a type alias would help readability a lot as well in my opinion.

Reply:
done